### PR TITLE
ECC unit test setup

### PR DESCRIPTION
**Pull request by cgsandford**
Replaces cubes input to ECC unit tests with cubes that have the expected IMPROVER metadata.  Changes to unit tests as follows:

- Stores shared data arrays for ECC unit tests in a new `ecc_test_data` module
- Replaces all calls to `ensemble_calibration.helper_functions` setup functions to either the `set_up_variable_cubes` module or the `build_spotdata_cube` function (via a local setup utility in `ecc_test_data`)
- Deletion of some duplicated tests in `test_EnsembleReordering` **only** - no changes to number / nature of tests in other modules
- Removal of unit test setup functions that are no longer used

These changes are applied to a large number of lines of code, but are limited in scope, and should therefore be possible to review in this form.

In addition, the following very minor code changes:

- Bug fix for time coordinate promotion in ensemble reordering
- Refactoring of percentile cube creation function in ECC utilities

Further code changes (including potentially centralising the spot data test cube setup utility) will be done in a follow-on PR, so that this PR can focus on the limited set of changes listed above.

Testing:
 - [x] Ran tests and they passed OK

### CODE REVIEW

File: improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/ecc_test_data.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    [
+        [1.0, 0.9, 1.0, 0.8, 0.9, 0.5, 0.5, 0.2, 0.0],
+        [1.0, 0.5, 1.0, 0.5, 0.5, 0.3, 0.2, 0.0, 0.0],
+        [1.0, 0.2, 0.5, 0.2, 0.0, 0.1, 0.0, 0.0, 0.0],
+    ],
+    dtype=np.float32,
+)
+
+
+def set_up_spot_test_cube(type="realization"):
```
Comment (benfitzpatrick) [inline improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/ecc_test_data.py:81]: Codacy found an issue: [Redefining built-in 'type'](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5762379)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/ecc_test_data.py  5
         

Complexity decreasing per file
==============================
+ improver/ensemble_copula_coupling/utilities.py  -3
         

Clones removed
==============
+ improver_tests/calibration/ensemble_calibration/helper_functions.py  -4
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_EnsembleReordering.py  -35
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_utilities.py  -3
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ResamplePercentiles.py  -5
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ConvertLocationAndScaleParametersToPercentiles.py  -7
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ConvertProbabilitiesToPercentiles.py  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5762379&bid=18862153)

File: improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ResamplePercentiles.py [hunk 2]
```diff
@@ -42,11 +43,12 @@
 )
 from improver.utilities.warnings_handler import ManageWarnings
 
-from ...calibration.ensemble_calibration.helper_functions import (
-    add_forecast_reference_time_and_forecast_period,
-    set_up_cube,
-    set_up_spot_temperature_cube,
+from ...set_up_test_cubes import (
```
Comment (benfitzpatrick) [inline improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ResamplePercentiles.py:46]: Codacy found an issue: [Unused set_up_variable_cube imported from set_up_test_cubes](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5762379)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/ecc_test_data.py  5
         

Complexity decreasing per file
==============================
+ improver/ensemble_copula_coupling/utilities.py  -3
         

Clones removed
==============
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_EnsembleReordering.py  -35
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ConvertLocationAndScaleParametersToPercentiles.py  -7
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ResamplePercentiles.py  -5
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_utilities.py  -3
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ConvertProbabilitiesToPercentiles.py  -4
+ improver_tests/calibration/ensemble_calibration/helper_functions.py  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5762379&bid=18862153)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/ecc_test_data.py  5
         

Complexity decreasing per file
==============================
+ improver/ensemble_copula_coupling/utilities.py  -3
         

Clones removed
==============
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ConvertLocationAndScaleParametersToPercentiles.py  -7
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ConvertProbabilitiesToPercentiles.py  -4
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ResamplePercentiles.py  -5
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_EnsembleReordering.py  -35
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_utilities.py  -3
+ improver_tests/calibration/ensemble_calibration/helper_functions.py  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5762379&bid=18862153)

File: improver/ensemble_copula_coupling/utilities.py [hunk 3]
```diff
@@ -158,31 +158,27 @@ def create_cube_with_percentiles(percentiles, template_cube, cube_data, cube_uni
...
+            )
+        )
+        cubes.append(cube)
+    result = cubes.merge_cube()
 
-    metadata_dict = copy.deepcopy(template_cube.metadata._asdict())
-    result = iris.cube.Cube(cube_data, **metadata_dict)
+    # if percentile is single-valued, promote new coordinate to dimension
+    if len(percentiles) == 1:
+        result = iris.util.new_axis(result, result.coord("percentile"))
```
Comment (VictoriaLouiseS) [inline improver/ensemble_copula_coupling/utilities.py:175]: Please could you explain why percentiles are initially being created as AuxCoords and then being promoted if single-valued?

Comment (cgsandford) [reply improver/ensemble_copula_coupling/utilities.py:175]: Good question.  I don't know why the single-value percentile coordinate is required to be dimensional, but that's what this function is providing, and unit tests for the calling class fail if the coordinate is not promoted.  It looked to me like some significant changes to the code might result from addressing this, which I didn't want to add here to an already large PR.  So I'm planning to address this and other issues that might require re-factoring in a follow-on PR.

I'll email you the link to the JIRA ticket on this, I'm planning to collate any outstanding review comments there to address in a follow-on PR.

File: improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_EnsembleReordering.py [hunk 4]
```diff
@@ -328,21 +191,17 @@ def test_ordered_data(self):
...
-                [[[3, 3, 3], [3, 3, 3], [3, 3, 3]]],
+                [[1, 1, 1], [1, 1, 1], [1, 1, 1]],
+                [[2, 2, 2], [2, 2, 2], [2, 2, 2]],
+                [[3, 3, 3], [3, 3, 3], [3, 3, 3]],
             ]
         )
 
         calibrated_data = raw_data
+        raw_cube = self.cube.copy(data=raw_data)
+        calibrated_cube = self.cube.copy(data=calibrated_data)
```
Comment (bayliffe) [inline improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_EnsembleReordering.py:202]: Why are we setting ```calibrated_data = raw_data ``` and then using it in this way? Just use ```raw_data```.

File: improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ConvertProbabilitiesToPercentiles.py [hunk 5]
```diff
@@ -173,57 +156,42 @@ class Test__probabilities_to_percentiles(IrisTest):
...
+            self.cube, self.percentiles, self.bounds_pairing
+        )
         self.assertIsInstance(result, Cube)
+        self.assertEqual(result.name(), "air_temperature")
 
     def test_unknown_thresholding(self):
         """Test that the plugin returns an Iris.cube.Cube."""
-        cube = self.current_temperature_forecast_cube
-        cube.coord(var_name="threshold").attributes[
+        self.cube.coord(var_name="threshold").attributes[
```
Comment (bayliffe) [inline improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ConvertProbabilitiesToPercentiles.py:177]: This doc-string is incorrect.

File: improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_EnsembleReordering.py (hunk 4 above)
Comment (cgsandford) [reply improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_EnsembleReordering.py:202]: OK

File: improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ConvertProbabilitiesToPercentiles.py (hunk 5 above)
Comment (cgsandford) [reply improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ConvertProbabilitiesToPercentiles.py:177]: Good spot...

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/ecc_test_data.py  5
         

Complexity decreasing per file
==============================
+ improver/ensemble_copula_coupling/utilities.py  -3
         

Clones removed
==============
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_EnsembleReordering.py  -35
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ResamplePercentiles.py  -5
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ConvertProbabilitiesToPercentiles.py  -4
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_utilities.py  -3
+ improver_tests/calibration/ensemble_calibration/helper_functions.py  -4
+ improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_ConvertLocationAndScaleParametersToPercentiles.py  -7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5762379&bid=18862153)

**Pull request merged**